optaplanner-core/src/main/java/org/optaplanner/core/impl/phase/AbstractPhase.java

```diff
@@ -53,7 +47,6 @@
 
     protected final int phaseIndex;
     protected final String logIndentation;
-    protected final Map<Tags, List<AtomicReference<Number>>> stepScoreMap = new ConcurrentHashMap<>();
```

Comment:
I checked the code and I'm not really sure why we need the `AtomicReference` there.
The `Number` is there because we do not know what the `Score` will return - fair enough. But the `AtomicReference`?

If you need concurrent modifications to the list, there is the `CopyOnWriteArrayList`.
Is there any particular reason why the change of the list element should be atomic? `List.set(...)` should be enough here.

Reply:
It more complex to do it with a list, since the Gauge function is given whatever is passed to it, meaning we would need to also store the index in the list to fetch versus just the atomic reference (currently, the atomic reference is passed to the gauge, and the gauge read directly from the atomic reference, bypassing the List). See Section 9 in https://micrometer.io/docs/concepts . In theory, it does not need to be atomic, just settable. However, the concept page only show atomic types.